### PR TITLE
Interactions schema

### PR DESCRIPTION
**Pull request by vinnyjth**
## DESCRIPTION

Companion PR, to be merged after: https://github.com/ApollosProject/apollos-templates/pull/2

### What does this PR do, or why is it needed?

Adds a `mutation` to create an `interaction` against a generic NodeId. That NodeId can be any type of Entity, but it will work best if it's a Rock Entity (for now, anyway). 

### What design trade-offs/decisions were made?

I cleaned up a lot of older code to hopefully make the purpose of this code more clear. 

### How do I test this PR?
1. Boot this PR in apollos-templates master
2. Login with this mutation
```
mutation login {
  authenticate(identity:"vincent@differential.com", password:"password") {
    token
  }
}
```
(store your token inside the Authorization header)
3. Run this mutation in graphiql
```
mutation {
  interactWithNode(nodeId:"WeekendContentItem:2b32682bd0d96179e6f001060251c96f", action: COMPLETE){
    success
    node {
      id
    }
  }
}
```
4. You should see your mutation show up here:
https://apollosrock.newspring.cc/page/510?ComponentId=263

### What automated tests did you write?

Tests that cover new functionality, and edited older tests to better cover all functionality.

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [x] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [x] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by vinnyjth** [comment]
To whomever reviews this: 

1. I hope to improve the caching using the RedisCache in a further PR.
2. There's an opportunity to eliminate the `createContentItemInteraction` and replace it with the `createNodeInteraction`. I need to peek at how the churches are using RockAnalytics before I do this.

**Comment by conrad-vanl** [comment]
It appears this currently only works with `contentItem`s? 

thoughts on _always_ storing our own Apollos ID as the foreign ID on Rock? And selectively _also_ storing an `entityType` / `entityId` if we detect that the `node` is a Rock entity?

**Comment by vinnyjth** [comment]
> It appears this currently only works with contentItems?

It only work's with Node's that that are backed by Rock Entities (I can try and come up with another type of node to demo)

> thoughts on always storing our own Apollos ID as the foreign ID on Rock? And selectively also storing an entityType / entityId if we detect that the node is a Rock entity?

That's a great idea! I'll introduce that in this PR.

**Comment by conrad-vanl** [comment]
> It only work's with Node's that that are backed by Rock Entities (I can try and come up with another type of node to demo)

I probably missed it - can you point out the logic where it determines the type of Rock Entity?

**Comment by vinnyjth** [comment]
@conrad-vanl https://github.com/ApollosProject/apollos-apps/pull/1374/files#diff-c5a860f49f5b9d28d391cdbf198599eeR44

**Comment by vinnyjth** [comment]
@conrad-vanl here's a non ContentItem mutation for you.
```
mutation {
  interactWithNode(nodeId: "Campus:559b23fd0aa90e81b1c023e72e230fa1", action:VIEW) {
    success
    node {
      id
    }
  }
}
```

This PR now uses the `foreignKey` field. I tried to use the foreignId field, but it only accepts `ints` and the `foreignGuid` only accepts actual guids.

**Comment by vinnyjth** [comment]
Added a detail to the PR description
---
Companion PR, to be merged after: ApollosProject/apollos-templates#2

**Comment by vinnyjth** [comment]
@conrad-vanl Anything else you want to see in this PR before I move forward? This is a key building block for the Next Step PR

File: packages/apollos-data-connector-rock/src/interactions/data-source.js [hunk 1]
```diff
@@ -34,4 +34,42 @@ export default class Interactions extends RockApolloDataSource {
...
+    const {
+      dataSources: { RockConstants, Auth },
+    } = this.context;
+    const { id, __type } = parseGlobalId(nodeId);
+
+    const entityType = await RockConstants.modelType(__type);
+
+    if (!entityType) {
+      console.error(
+        'nodeId is an invalid (non-rock) entity type. This is not yet supported.'
```
Comment (conrad-vanl) [inline packages/apollos-data-connector-rock/src/interactions/data-source.js]: Would supporting a non-rock entity type be as simple as not setting the `entity` fields below? Would it be "easy" to not make Rock entities a requirement with this PR? Or is that larger work you're thinking you'll follow up with?

File: packages/apollos-data-connector-rock/src/rock-constants/index.js [hunk 2]
```diff
@@ -49,7 +50,7 @@ class RockConstants extends RockApolloDataSource {
       model: 'InteractionChannels',
       objectAttributes: {
         Name: channelName,
-        UsesSession: true,
+        UsesSession: false,
```
Comment (conrad-vanl) [inline packages/apollos-data-connector-rock/src/rock-constants/index.js]: What does changing this to false do?

File: packages/apollos-data-schema/src/index.js [hunk 3]
```diff
@@ -250,6 +250,33 @@ export const analyticsSchema = gql`
   }
 `;
 
+export const interactionsSchema = gql`
+  scalar InteractionValue
+
+  enum InteractionAction {
+    VIEW
+    COMPETE
```
Comment (conrad-vanl) [inline packages/apollos-data-schema/src/index.js]: Typo?

Comment (conrad-vanl) [reply packages/apollos-data-schema/src/index.js]: Also, should we keep defined interactions super small for now?

Can church apps extend enums and add more? Like if I wanted to track a different action in my app that's not on core, say PRAYED - how would I do that?

File: scripts/watch.js [hunk 4]
```diff
@@ -22,7 +22,7 @@ chokidar
         console.log(stdout, stderr);
         console.log(`yarn lerna --scope ${pkg} exec -- yalc push`);
         exec(
-          `./node_modules/lerna/cli.js --scope @apollosproject/data-connector-twilio exec -- yalc push --changed`,
+          `./node_modules/lerna/cli.js --scope ${pkg} exec -- yalc push --changed`,
```
Comment (conrad-vanl) [inline scripts/watch.js]: How is this change related?

File: packages/apollos-data-connector-rock/src/interactions/data-source.js (hunk 1 above)
Comment (vinnyjth) [reply packages/apollos-data-connector-rock/src/interactions/data-source.js]: I it will be that, plus a change to the `interactionComponent` and `interactionChannel` method so they handle a `null entityType` state fine. I wanted to tackle that in a future PR (and I'll add it to the todo list) I would like to keep this PR as focused as a I can, and then dow a follow up with non-rock entities.

File: scripts/watch.js (hunk 4 above)
Comment (vinnyjth) [reply scripts/watch.js]: It's not - it fixes a bug when working with the API where any change you make only pulls in the `twilio` package 🙃. It's a typo from when I moved us to `yalc`.

File: packages/apollos-data-schema/src/index.js (hunk 3 above)
Comment (vinnyjth) [reply packages/apollos-data-schema/src/index.js]: Whoops! I had it fixed locally and forgot to push. 

I'm all for keeping defined interactions super small. Extending them stinks, you have to copy over the schema and change it, which isn't super hard but it's frustrating. 

That said, in graphql 15.0 you can `extend InteractionAction`. We haven't updated to that version yet, but I think we can whenever we take the time to do it.

File: packages/apollos-data-connector-rock/src/rock-constants/index.js (hunk 2 above)
Comment (vinnyjth) [reply packages/apollos-data-connector-rock/src/rock-constants/index.js]: It does two things 
1. Stops us from needing to keep track of sessionIds (which we are currently storing in the user token) 
2. Changes the UI in Rock from being oriented towards "what users have done what" to "what things have had what done to them" 

I don't have a good argument for going one way or or the other. I would say that `UsesSessions: true` will make it harder for us to create actions for anonymous users.

**Comment by vinnyjth** [comment]
@conrad-vanl Are there more changes you want me to make here?

File: packages/apollos-data-schema/src/index.js (hunk 3 above)
Comment (conrad-vanl) [reply packages/apollos-data-schema/src/index.js]: ok, lets add upgrade to graphql 15.0 to the todos. We get to keep this as an `enum` right?

Comment (vinnyjth) [reply packages/apollos-data-schema/src/index.js]: `enum` or `string`. I'll prioritize that 15.0 upgrade
